src/backend/kvm/builder.rs

```diff
@@ -0,0 +1,111 @@
+// SPDX-License-Identifier: Apache-2.0
+
+use super::config::Config;
+
+use std::convert::TryFrom;
+use std::sync::{Arc, RwLock};
+
+use crate::backend::kvm::mem::Region;
+use anyhow::{Error, Result};
+use kvm_bindings::bindings::kvm_userspace_memory_region;
+use kvm_bindings::fam_wrappers::KVM_MAX_CPUID_ENTRIES;
+use kvm_ioctls::{Kvm, VmFd};
+use lset::Span;
+use mmarinus::{perms, Map};
+use sallyport::elf::pf::kvm::SALLYPORT;
+use sallyport::Block;
+use std::collections::VecDeque;
+use std::mem::size_of;
+use std::num::NonZeroUsize;
+use x86_64::VirtAddr;
+
+pub struct Builder {
+    _cnfg: Config,
+    kvm_fd: Kvm,
+    vm_fd: VmFd,
+    regions: Vec<Region>,
+    sallyport_blocks: Option<Span<VirtAddr, NonZeroUsize>>,
+}
+
+impl TryFrom<super::config::Config> for Builder {
+    type Error = Error;
+
+    fn try_from(config: super::config::Config) -> Result<Self> {
+        let kvm_fd = Kvm::new().unwrap();
+        let vm_fd = kvm_fd.create_vm().unwrap();
+        vm_fd.create_irq_chip().unwrap();
+
+        Ok(Builder {
+            kvm_fd,
+            vm_fd,
+            regions: Vec::new(),
+            _cnfg: config,
+            sallyport_blocks: None,
+        })
+    }
+}
+
+impl super::super::Mapper for Builder {
+    type Config = super::config::Config;
+    type Output = Arc<dyn super::super::Keep>;
+
+    fn map(&mut self, pages: Map<perms::ReadWrite>, to: usize, with: u32) -> anyhow::Result<()> {
+        // Ignore regions with no pages.
+        if pages.is_empty() {
+            return Ok(());
+        }
+
+        if with & SALLYPORT != 0 {
+            let sallyport_blocks = Span {
+                start: VirtAddr::from_ptr(pages.as_ptr()),
+                count: NonZeroUsize::new(pages.size() / size_of::<Block>()).unwrap(),
+            };
+            self.sallyport_blocks.replace(sallyport_blocks);
+        }
+
+        let mem_region = kvm_userspace_memory_region {
+            slot: self.regions.len() as _,
+            flags: 0,
+            guest_phys_addr: to as _,
+            memory_size: pages.size() as _,
+            userspace_addr: pages.addr() as _,
+        };
+
+        unsafe {
+            self.vm_fd.set_user_memory_region(mem_region).unwrap();
+        }
```

Comment:
`unsafe { self.vm_fd.set_user_memory_region(mem_region).unwrap() };`

If you put the semicolon after the unsafe block then `cargo fmt` won't wrap it.

Reply:
yes